### ISSUE TITLE
false positives

### ISSUE THREAD

**Issue by wilson9x1** [stamparm/DSXS#5]
good project!!
but, find this rule will cause false positives:
```
(r"<[^>]*'[^>']*%(chars)s[^>']*'[^>]*>", ('\'',), "\"<.'.xss.'.>\", inside the tag, inside single-quotes, %(filtering)s filtering", r"(?s)<script.+?</script>|<!--.*?-->") 
```
egg:
demo.php:
```
<?php
$input= str_replace("\"", "&quot;" ,$_GET['xss']);
$input=str_replace(">","&lt;",$input);
$input=str_replace("<","&gt;",$input);
echo '<meta name="description" content="';
print($input);
echo '#23578';
print($input);
echo '#23578';
print($input);
echo '#23578/>';
?>
```

![2018-01-09 3 06 48](https://user-images.githubusercontent.com/9835472/34708673-c668a864-f54e-11e7-8696-b1d13654f9d8.png)

for help～

**Comment by wilson9x1** [comment]
Another false positives:

demo:
```
</main><script>
    var jobs = {
        log: {
            pj: 'jobs-social',
            qid: '',
            pn: 0
        },
        common: {
			query:'',
            listNum:'0',
            dispNum:'0',
            loc:'aaaa',
			pn: 0,
			tid :'' || '' || 'home',
			qid :'2388629235',
            asyncUrl: '/api/quanzhiasync',
            jumpUrl: '/szzw',
            cookieStr: 'URLTITLESALARY',
			selected : {"city":"<?php echo str_replace("\"", "\\\"" ,$_GET['xss']);?>"}
		}
    };
</script>
```

**Comment by stamparm** [comment]
This is a project under 100 lines which should cover majority of XSS vulnerabilities. It doesn't have a true-positive or false-positive detection mechanism, neither it could be implemented into such restricted code environment.

I've personally made another detection mechanism which is more complex than this one and I am aware that DSXS has its quirks. Though, as a PoC I would say that it works pretty fine. It is up to the tester to test the positiveness of the results (as in any scanner/tool)

**Comment by stamparm** [comment]
With the latest patch problem related to the https://github.com/stamparm/DSXS/issues/5#issuecomment-356805571 should be gone

**Comment by stamparm** [comment]
With the latest patch problem related to the https://github.com/stamparm/DSXS/issues/5#issue-286990471 should be gone, though...

Now the new problem arise where another rule is being triggered because the original content is actually being reflected in an invalid matter (unclosed double quotes) which is something unexpected. It causes (minor) problem where everything that looks like a quoted string is being prefiltered (before the rule is being triggered).

**Comment by wilson9x1** [comment]
oooo.....  thank you for answer~~

**Comment by phackt** [comment]
@stamparm Hello Miroslav, you say it s a POC but i did not find any more convenient tool. It s pretty quick and effective. However you were talking about another of your tools with an improved detection mechanism, is it a private project or is there a repo for this ? Thanks.

**Comment by stamparm** [comment]
@phackt private project, developed in several occasions (cca. two months of development time). Greasemonkey script created for automated vulnerability assessment (with similar (though improved) web application security heuristics as in DSXS or DSSS or...) during regular browsing of web pages. Now the "fishing story" goes like this - it got too dangerous to release it wide open to the public as I got great results by just browsing random sites. In one wild scenario I wanted to release it on Defcon, though I got rejected (based just on abstract) two years ago.

In short, private project currently not releasable to anyone as it would produce new wave of script kiddies (hack as you browse).

**Comment by phackt** [comment]
@stamparm Arf at least you have a rocking tool for your hunt. I will look at your code i just would like to provide the payload in this way (http://mysite.com/<payload>), let's say if no params have been found in the url (could be useful for xss in tracker and co). Cheers.

**Comment by phackt** [comment]
@wilson9x1 Hi Wilson, did you find any new regexp, i'm also facing some false/positives with the regexp: ```r"(?s)<script[^>]*>[^<]*?'[^<']*%(chars)s|%(chars)s[^<']*'[^<]*</script>```.

For example for this payload: ```<script>var tutu="<test>"; var test="tutu\"";</script>```, it goes to the last 'script tag' regexp and triggers a vuln. I was wondering why the first both regex inside script tags enclosed by quotes do not want to look for '<', example for ```r"(?s)<script[^>]*>[^<]*?'[^<']*%(chars)s|%(chars)s[^<']*'[^<]*</script>```. I tried an other more permissive and non greedy regexp but it was not working: ```r'(?s)<script[^>]*>.*?".*?%(chars)s|%(chars)s.*?".*?</script>'```

Think i'm missing something, pretty complex to match the whole thing without false/positives. If you guys have any idea. 
Cheers.

**Comment by phackt** [comment]
Ok @wilson9x1 i tried this kind of regexp ```(?s)<script[^>]*>[^<]*?"[^<"\']*%(chars)s|%(chars)s[^<"]*"[^<]*</script>```, but it does not trigger your example (escaped double quote), pretty weird, so it s caught by this one ```(?s)<script[^>]*>[^<]*?%(chars)s|%(chars)s[^<]*</script>``` and triggers a false positive
